resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -340,6 +340,30 @@ private[spark] class Client(
     amContainer.setTokens(ByteBuffer.wrap(serializedCreds))
   }
 
+  /**
+   * Set configurations sent from AM to RM for renewing delegation tokens.
+   */
+  private def setTokenConf(amContainer: ContainerLaunchContext): Unit = {
+    // SPARK-37205: this regex is used to grep a list of configurations and send them to YARN RM
+    // for fetching delegation tokens. See YARN-5910 for more details.
+    // The feature is only supported in Hadoop 3.x and up, hence the check below.
+    val regex = sparkConf.get(config.AM_SEND_TOKEN_CONF)
+    if (regex != null && regex.nonEmpty && VersionUtils.isHadoop3) {
+      logInfo(s"Processing token conf (spark.yarn.am.sendTokenConf) with regex $regex")
+      val dob = new DataOutputBuffer();
+      val copy = new Configuration(false);
+      copy.clear();
+      hadoopConf.asScala.foreach { entry =>
+        if (entry.getKey.matches(regex)) {
+          copy.set(entry.getKey, entry.getValue)
+          logInfo(s"Captured key: ${entry.getKey} -> value: ${entry.getValue}")
+        }
+      }
+      copy.write(dob);
+      amContainer.setTokensConf(ByteBuffer.wrap(dob.getData))
```

Comment:
Just a question. The compilation works with Hadoop 2.7, right?

Reply:
Oops, it won't work for Hadoop 2.7. Hmm let me think how to make it work ..

Reply:
I guess we can follow `ResourceRequestHelper` and use reflection to lookup the method when using Hadoop 3.x, to avoid compilation error.